### ISSUE TITLE
Context length

### ISSUE THREAD

**Issue by skyne98** [apage43/mptgen#1]
Hey, hey, just here to say - awesome project! Forked it and made it work on my side too, hoping to make it work using WASM and on mobile, as I managed to use the quantized 4-bit model 👀  I hope you don't mind!

I am not certain how to implement the context adjustable context length you mention (and which the model supports), so I wanted to encourage you! Do you think it's fairly easy to do? 

Also, what should the drawbacks of a long 65k+ context window be? Longer processing times?

Thanks a lot!

**Comment by apage43** [comment]
tl;dr basic support (except StoryWriter) is not hard but it will require more ram and compute to work with longer inputs, it will get slower as the input gets longer.

--

To just allow using larger context with the currently supported models mainly means making the storage size of the attention cache (`memory_k`, `memory_v`) bigger https://github.com/apage43/mptgen/blob/b744c3a71fe2037bc6c1e94b0a303eb402877333/minmpt.cpp/mpt.cpp#L181-L182 
The easy solution here is just to allow overriding the `n_ctx` parameter loaded from the model file before its used to calculate the storage size (llama.cpp already lets you do this, for example), probably with an extra arg to mpt_model_load. It should also be technically *possible* to resize/reallocate these after the model's been loaded, but properly copying the data around to make it possible to continue your current session could get tricky.

Also, while you can can *run* these kinds of models on a sequence of any size you'd like (most of them just don't work very well on inputs longer than they were trained on, MPTs should be a bit better at that), the amount of memory and compute you need to do so increases significantly with the context length. Memory to store the attention cache will increase linearly, but the amount of scratch memory needed to *evaluate the model itself* will increase *quadratically* because of the attention operation.

Supporting StoryWriter is also a bit more of a lift because it requires a couple additional tweaks to the model code, as it makes use of `clip_qkv` and a non-default value of `alibi_bias_max` https://huggingface.co/mosaicml/mpt-7b-storywriter/blob/main/config.json#L7-L12

* `clip_qkv` would go here https://huggingface.co/mosaicml/mpt-7b-storywriter/blob/main/config.json#L7-L12 and would only need to clamp the values in `cur` to + or - `clip_qkv`, but currently ggml doesn't have a built in `clamp` operation
* `alibi_bias_max` replaces the 8 (and 4) here: https://github.com/apage43/mptgen/blob/b744c3a71fe2037bc6c1e94b0a303eb402877333/minmpt.cpp/ggml.c#L8404 so supporting it means adding a way to pass that through the `ggml_alibi` operator.

Probably also should make sure to support building ggml with BLAS support for big matmuls, which I haven't yet copied over, but will likely help speed at bigger context lengths.

**Comment by skyne98** [comment]
First of all, thanks for such a quick answer!

1. About the memory usage - if I get it correctly, all of the increases you described relate to either fairly small amounts in the first place (attention cache) or moment to moment allocations, which are at any moment in time by themselves not huge, right? I mean, comparatively to all the weights we store in memory already.

2. About the BLAS, yeah, I have a RWKV runtime I have been working on, which is an RNN based LLM, with very promising results (also promises "infinite" context length due to the architecture, though, RNNs are known to "blur" their context the further it is away from the current iteration). There in that runtime I have put lots of effort on CPU optimization, but, surprisingly, MPT7B in GGML, even without BLAS, is pretty fast!

Also, what are your thoughts on RWKV? Maybe you have heard about it already?

**Comment by apage43** [comment]
> moment to moment allocations

ah, one thing about ggml currently is that it doesn't alloc or free *during* the actual computation - you allocate a slab of memory big enough to handle all the intermediate values in the whole graph and only free it at the end (most ggml examples put it in a static variable and never free it at all, they'll resize it to be bigger as the context length increases, but never shrink it) - right now there's not a good way of reusing the memory the previous layer had been using so you wind up keeping values from all layers until the end - its not as bad as it could be as many operations are done in-place but it does use more memory than it really *has* to

have heard about rwkv but have not actually tried using it yet

**Comment by skyne98** [comment]
Yeah, I am currently stuck between choosing where to invest time and effort. :D RWKV feels a little rough currently, with its creator just being one person backed by EAI and the model being unpredictable occasionally at certain tasks. However, it does, basically, support "infinite" context lengths with no significant increases in resources, just due to its RNN nature. Also, the whole runtime for it can be implemented from 0 in ~200 lines of Rust without almost any dependencies. Will be exciting to have a person, who knows what they are talking about, have a look at this!

**Comment by apage43** [comment]
there's now a `--n-ctx` flag on the example bins and support for overriding it in the bindings
